src/libvfs.c

```diff
@@ -1,125 +1,231 @@
+  METHOD(xRead);
+  METHOD(xWrite);
+  METHOD(xTruncate);
+  METHOD(xSync);
+  METHOD(xFileSize);
+  METHOD(xLock);
+  METHOD(xUnlock);
+  METHOD(xCheckReservedLock);
+  METHOD(xFileControl);
+  METHOD(xSectorSize);
+  METHOD(xDeviceCharacteristics);
+  METHOD(xShmMap);
+  METHOD(xShmLock);
+  METHOD(xShmBarrier);
+  METHOD(xShmUnmap);
+#undef METHOD
+  pFile->pMethods = pMethods;
+  ((VFSFile*)pFile)->pVfs = pVfsExt;
   return result;
 }
 
-void *EMSCRIPTEN_KEEPALIVE getSqliteFree()
-{
-  return sqlite3_free;
+static int libvfs_xDelete(sqlite3_vfs* pVfs, const char* zName, int syncDir) {
+  return VFS_JS(ippppi, pVfs, xDelete, pVfs, zName, syncDir);
 }
 
-int main()
-{
-  sqlite3_initialize();
-  return 0;
+static int libvfs_xAccess(sqlite3_vfs* pVfs, const char* zName, int flags, int* pResOut) {
+  return VFS_JS(ippppip, pVfs, xAccess, pVfs, zName, flags, pResOut);
+}
+
+static int libvfs_xFullPathname(sqlite3_vfs* pVfs, const char* zName, int nOut, char* zOut) {
+  return VFS_JS(ippppip, pVfs, xFullPathname, pVfs, zName, nOut, zOut);
+}
+
+static int libvfs_xRandomness(sqlite3_vfs* pVfs, int nBuf, char* zBuf) {
+  return VFS_JS(ipppip, pVfs, xRandomness, pVfs, nBuf, zBuf);
+}
+
+static int libvfs_xSleep(sqlite3_vfs* pVfs, int microseconds) {
+  return VFS_JS(ipppi, pVfs, xSleep, pVfs, microseconds);
+}
+
+static int libvfs_xCurrentTime(sqlite3_vfs* pVfs, double* pJulianDay) {
+  return VFS_JS(ipppp, pVfs, xCurrentTime, pVfs, pJulianDay);
+}
+
+static int libvfs_xGetLastError(sqlite3_vfs* pVfs, int nBuf, char* zBuf) {
+  return VFS_JS(ipppip, pVfs, xGetLastError, pVfs, nBuf, zBuf);
+}
+
+static int libvfs_xCurrentTimeInt64(sqlite3_vfs* pVfs, sqlite3_int64* pTime) {
+  return VFS_JS(ipppp, pVfs, xCurrentTimeInt64, pVfs, pTime);
+}
+
+int EMSCRIPTEN_KEEPALIVE libvfs_vfs_register(
+  const char* zName,
+  int mxPathName,
+  int methodMask,
+  int asyncMask,
+  int makeDefault,
+  void** ppVfs) {
+  // Get the current default VFS to use if methods are not defined.
+  const sqlite3_vfs* backupVfs = sqlite3_vfs_find(NULL);
+
+  // Allocate and populate the new VFS.
+  VFS* vfs = (VFS*)sqlite3_malloc(sizeof(VFS));
+  if (!vfs) return SQLITE_NOMEM;
+  bzero(vfs, sizeof(VFS));
+
+  vfs->base.iVersion = 2;
+  vfs->base.szOsFile = sizeof(VFSFile);
+  vfs->base.mxPathname = mxPathName;
+  vfs->base.zName = strdup(zName);
+
+  // The VFS methods go to the adapter implementations in this file,
+  // or to the default VFS if the JavaScript method is not defined.
+#define METHOD(NAME) vfs->base.NAME = \
+  (methodMask & (1 << NAME)) ? libvfs_##NAME : backupVfs->NAME
+
+  METHOD(xOpen);
+  METHOD(xDelete);
+  METHOD(xAccess);
+  METHOD(xFullPathname);
+  METHOD(xRandomness);
+  METHOD(xSleep);
+  METHOD(xCurrentTime);
+  METHOD(xGetLastError);
+  METHOD(xCurrentTimeInt64);
+#undef METHOD
+
+  vfs->methodMask = methodMask;
+  vfs->asyncMask = asyncMask;
+
+  *ppVfs = vfs;
+  return sqlite3_vfs_register(&vfs->base, makeDefault);
 }
 
 int setup_powersync()
```

Comment:
Not directly related to this PR, but can the powersync setup be moved to main.c? libvfs.c is a weird place for it

Reply:
Agreed. Will update.